### PR TITLE
feat(protocol): add IPreconfViolationVerifier for preconfirmation

### PR DESCRIPTION
**Pull request by dantaik**
Introduce an interface to support proving various types of preconfirmation violations. I expect Nethermind's registry can be designed to enable multiple `IPreconfViolationProver` implementations, allowing us and others to implement our own preconfirmation commitment without tightly coupling with Nethermind's code.

Please do not merge!

### CODE REVIEW

**Review by davidtaikocha — APPROVED** [review]

File: packages/protocol/contracts/layer1/verifiers/SgxVerifierBase.sol [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            instances[nextInstanceId] = Instance(_instances[i], validSince);
+            ids[i] = nextInstanceId;
+
+            emit InstanceAdded(nextInstanceId, _instances[i], address(0), validSince);
+
+            ++nextInstanceId;
+        }
+    }
+
+    function _replaceInstance(uint256 id, address oldInstance, address newInstance) internal {
```
Comment (swarna1101) [inline packages/protocol/contracts/layer1/verifiers/SgxVerifierBase.sol:161]: Can we cache the instance in a local variable, we can minimize storage reads then here in `_replaceInstance`

File: packages/protocol/contracts/layer1/preconf/PreconfViolationSgxVerifier.sol [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                _receipt.chainId,
+                _receipt.blockId,
+                _receipt.position,
+                _receipt.isNonRevertGuaranteed,
+                _receipt.preconfer
+            )
+        );
+    }
+
+    function taikoChainId() internal view virtual returns (uint64) {
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/layer1/preconf/PreconfViolationSgxVerifier.sol]: See RollupAddressCache and SharedAddressCache

File: packages/protocol/contracts/layer1/verifiers/SgxVerifierBase.sol [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+            if (instances[idx].addr == address(0)) revert SGX_INVALID_INSTANCE();
+
+            emit InstanceDeleted(idx, instances[idx].addr);
+
+            delete instances[idx];
+        }
+    }
+
+    function _addInstances(
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/layer1/verifiers/SgxVerifierBase.sol:130]: In the newer versions of Solidity (starting from 0.8.0), ++i inside a for loop does not require explicit wrapping in unchecked. The Solidity compiler automatically applies unchecked arithmetic in for loop increments and decrements, meaning overflow/underflow checks are not performed for the increment/decrement operation by default.

File: packages/protocol/contracts/layer1/verifiers/SgxVerifierBase.sol (hunk 1 above)
Comment (dantaik) [reply packages/protocol/contracts/layer1/verifiers/SgxVerifierBase.sol:161]: Sorry which read?

**Comment by dantaik** [comment]
This PR is closed and all changes were moved to https://github.com/taikoxyz/taiko-mono/pull/18212

**Pull request closed without merging**